File: cpp/july/day_28_Task Scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int> counts(26, 0);
        int i, j, total = 0, curr, cycle = n+1;
        for(i = 0; i < tasks.size(); i++){
            counts[tasks[i] - 'A'] += 1;
        }
        priority_queue<int> pq;
        for(i = 0; i < 26; i++){
            if(counts[i]){
                pq.push(counts[i]);
            }
        }
        while(!pq.empty()){
            vector<int> tmp;
            curr = 0;
            for(i = 0; i < cycle; i++){
                if(!pq.empty()){
                    tmp.push_back(pq.top());
                    pq.pop();
                    curr += 1;
                }
            }
            
            for(i = 0; i < tmp.size(); i++){
                tmp[i]--;
                if(tmp[i]){
                    pq.push(tmp[i]);
                }
            }
            total += !pq.empty() ? cycle : curr;
        }
        return total;
    }
};

File: python3/july/day_28_Task Scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        count = [0] * 26
        max_count = 0
        for task in tasks:
            c = int(ord(task) - ord('A'))
            count[c] += 1
            max_count = max(max_count, count[c])
        
        ret = (max_count - 1) * (n + 1)
        for i in range(26):
            if (count[i] == max_count):
                ret += 1

        return max(len(tasks), ret)
